### PR TITLE
adding new top level command account

### PR DESCRIPTION
**Pull request by ghost**
Adding part of the functionalities requested: [JIRA TICKET](https://jira.coreos.com/browse/SREP-1579)
```bash
$ uhc account roles
Resource
Subscription
Cluster
...
```

```bash
$ uhc account roles <roleName>
{
  id: "role_id"
  name: "role_name"
  ...
}
```
```bash
$ uhc account users
<list of users and their roles in my org>
```

```bash
$ uhc account users --org <orgId>
<list of users and their roles in a specific org> 
```

### CODE REVIEW

File: cmd/uhc/account/roles/cmd.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+		response, err := rolesListRequest.Send()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Can't send request: %v\n", err)
+			os.Exit(1)
+		}
+		response.Items().Each(func(item *amv1.Role) bool {
+			fmt.Println(item.ID())
+			return true
+		})
+		// Role name was provided.
```
Comment (fahlmant) [inline cmd/uhc/account/roles/cmd.go]: I think this comment is misplaced

File: cmd/uhc/account/users/cmd.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+		}
+		userOrg, _ := userConn.Body().GetOrganization()
+		args.org = userOrg.ID()
+	}
+
+	// Print top.
+	fmt.Println(stringPad("USER", namePad), "ROLES")
+	fmt.Println()
+
+	// Display a list of all users in our organization and tehir roles:
```
Comment (fahlmant) [inline cmd/uhc/account/users/cmd.go]: s/tehir/their/

File: cmd/uhc/account/cmd.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/openshift-online/uhc-cli/cmd/uhc/account/quota"
+	"github.com/openshift-online/uhc-cli/cmd/uhc/account/roles"
+	"github.com/openshift-online/uhc-cli/cmd/uhc/account/status"
+	"github.com/openshift-online/uhc-cli/cmd/uhc/account/users"
+	"github.com/spf13/cobra"
```
Comment (jhernand) [inline cmd/uhc/account/cmd.go]: Remember the order of imports: first standard library, then other projects, then this project.

File: cmd/uhc/account/quota/cmd.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	"strings"
+
+	"github.com/openshift-online/uhc-cli/pkg/dump"
+
+	"github.com/openshift-online/uhc-cli/pkg/config"
+	"github.com/openshift-online/uhc-cli/pkg/util"
+	"github.com/openshift-online/uhc-sdk-go/pkg/client"
+	amv1 "github.com/openshift-online/uhc-sdk-go/pkg/client/accountsmgmt/v1"
+
+	"github.com/spf13/cobra"
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go]: Fix the order of imports.

File: cmd/uhc/account/quota/cmd.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+		Tokens(cfg.AccessToken, cfg.RefreshToken).
+		Insecure(cfg.Insecure).
+		Build()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Can't create connection: %v\n", err)
+		os.Exit(1)
+	}
+	defer connection.Close()
+
+	// // Organization to search in case one was not provided:
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go]: Remove the duplicated comment prefix.

File: cmd/uhc/account/quota/cmd.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	if args.org == "" {
+		// Get organization of current user:
+		userConn, err := connection.AccountsMgmt().V1().CurrentAccount().Get().
+			Send()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Can't retrieve current user information: %v\n", err)
+			os.Exit(1)
+		}
+		userOrg, _ := userConn.Body().GetOrganization()
+		args.org = userOrg.ID()
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go]: Modifying `args` is not good practice, in my opinion. It means that if later you will need to know what was the value provided by the user you will not be able to. I'd suggest that you use a different variable for that:

```go
orgID := args.org

if args.org == "" {
        ...
       orgID = userOrg.ID()
}
```

Then later you can use `args.org` or `orgID` as needed.

File: cmd/uhc/account/quota/cmd.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+			orgResponse.Body().Name(), orgResponse.Body().ID()))
+		fmt.Println()
+		quotasResponse.Items().Each(func(quota *amv1.ResourceQuota) bool {
+			fmt.Println(fmt.Sprintf("%s-AZ: %d/%d", strings.ToUpper(quota.AvailabilityZoneType()),
+				quota.Reserved(), quota.Allowed()))
+			return true
+		})
+		fmt.Println()
+		return
+	}
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go:138]: Add a blank line here.

File: cmd/uhc/account/quota/cmd.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+	quotaRequest := orgRequest.ResourceQuota()
+
+	// Simple output:
+	if !args.json {
+
+		// Request
+		quotasResponse, err := quotaRequest.List().
+			Send()
+
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go]: Remove this blank line.

File: cmd/uhc/account/quota/cmd.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to retrieve quota: %v\n", err)
+			os.Exit(1)
+		}
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go:125]: Add a blank line here. And remember to end comment lines with punctionation: `// Display quota information:`

File: cmd/uhc/account/quota/cmd.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+		// Display quota information
+		fmt.Println(fmt.Sprintf("Cluster quota for organization '%s' ID: '%s'",
+			orgResponse.Body().Name(), orgResponse.Body().ID()))
+		fmt.Println()
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go]: You can do this with one single `Printf`:

```go
fmt.Printf("Cluster quota ...\n", orgResponse.Body().Name(), ...)
```

File: cmd/uhc/account/quota/cmd.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+		quotasResponse.Items().Each(func(quota *amv1.ResourceQuota) bool {
+			fmt.Println(fmt.Sprintf("%s-AZ: %d/%d", strings.ToUpper(quota.AvailabilityZoneType()),
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go]: Same: no need for `Println` and `Sprintf`; a single `Printf` does the same.

File: cmd/uhc/account/quota/cmd.go [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+		os.Exit(1)
+	}
+	quotaRequest := orgRequest.ResourceQuota()
+
+	// Simple output:
+	if !args.json {
+
+		// Request
+		quotasResponse, err := quotaRequest.List().
+			Send()
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go:121]: Remember that here you will get only the first page of results. Is that what you want?

File: cmd/uhc/account/quota/cmd.go [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Can't retrieve current user information: %v\n", err)
+			os.Exit(1)
+		}
+		userOrg, _ := userConn.Body().GetOrganization()
+		args.org = userOrg.ID()
+	}
+
+	// Get connection
+	orgRequest := connection.AccountsMgmt().V1().Organizations().Organization(args.org)
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go]: The name `orgRequest` is a bit confusing. What you have here is not a request, but a client for the resource that manages the organization. For this kind of variables I'd suggest to use `orgResource` (or `orgCollection` when the resource is a collection).

File: cmd/uhc/account/quota/cmd.go [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+	orgResponse, err := orgRequest.Get().Send()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Can't retrieve organization information: %v\n", err)
+		os.Exit(1)
+	}
+	quotaRequest := orgRequest.ResourceQuota()
```
Comment (jhernand) [inline cmd/uhc/account/quota/cmd.go]: Same; this is not a request but a client for a resource.

File: cmd/uhc/account/roles/cmd.go [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+		Insecure(cfg.Insecure).
+		Build()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Can't create connection: %v\n", err)
+		os.Exit(1)
+	}
+	defer connection.Close()
+
+	// No role name was provided,
+	// Print all roles.
```
Comment (jhernand) [inline cmd/uhc/account/roles/cmd.go]: No need for two lines here, one is enough and more readable: `// No role was provided; print al the roles.

File: cmd/uhc/account/roles/cmd.go [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+	if len(argv) < 1 {
+		rolesListRequest := connection.AccountsMgmt().V1().Roles().List()
+		response, err := rolesListRequest.Send()
```
Comment (jhernand) [inline cmd/uhc/account/roles/cmd.go]: Remember that this returns only the first page of results.

File: cmd/uhc/account/roles/cmd.go [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+		// Get role with provided id response:
+		roleResponse, err := connection.AccountsMgmt().V1().Roles().Role(argv[0]).Get().
+			Send()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Can't send request: %v\n", err)
+			os.Exit(1)
+		}
+		role := roleResponse.Body()
+
+		// Use role in new get request since original provides incomplete data.
```
Comment (jhernand) [inline cmd/uhc/account/roles/cmd.go]: What data is missing? That needs to be fixed in the SDK.

**Comment by fahlmant** [comment]
[test]

File: cmd/uhc/account/users/cmd.go [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+		&args.debug,
+		"debug",
+		false,
+		"Enable debug mode.",
+	)
+	flags.StringVar(
+		&args.org,
+		"org",
+		"", // Default value gets assigned later as connection is needed.
+		"Organization ID, deafaults to current users organization.",
```
Comment (jhernand) [inline cmd/uhc/account/users/cmd.go]: There is a typo in `deafaults`. And a period is more grammatically correct here: `Organization identifier. Defaults to the organization of the current user.`

File: cmd/uhc/account/users/cmd.go [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+	// Organization to search in case one was not provided:
+	if args.org == "" {
+		// Get organization of current user:
+		userConn, err := connection.AccountsMgmt().V1().CurrentAccount().Get().
+			Send()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Can't retrieve current user information: %v\n", err)
+			os.Exit(1)
+		}
+		userOrg, _ := userConn.Body().GetOrganization()
```
Comment (jhernand) [inline cmd/uhc/account/users/cmd.go]: If you are going to ignore the second value returned by `GetOrganization()` then it is better to use just `Organization()`. But note that the returned value may be `nil`, so you can't do the next step without checking.

File: cmd/uhc/account/users/cmd.go [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+	fmt.Println(stringPad("USER", namePad), "ROLES")
+	fmt.Println()
+
+	// Display a list of all users in our organization and their roles:
+	for {
+		// Get list of accounts in organization.
+		usersConn := connection.AccountsMgmt().V1().Accounts().List().Size(pageSize).Page(pageIndex)
+
+		// Format search request:
+		searchRequest := ""
```
Comment (jhernand) [inline cmd/uhc/account/users/cmd.go]: This initialization is redundant, as you are immediately assigning another value.

Also the `...Request` suffix it as bit confusing, as there are other requests in this context. I'd say that this is rather a `searchQuery`.

I'd suggest that you first calculate this query, and then send the request without creating the intermediate request variable:

```go
// Format search query:
searchQuery := ...

// Get all the users within the organization:
usersResponse, err := connection.AccountsMgmt().V1().Accounts().List().
        Size(pageSize).
        Page(pageIndex).
        Search(searchQuery).
        Send()
if err != nil {
        ...
}
```

**Review by jhernand — CHANGES_REQUESTED** [review]
Thaks @ALimaRedHat. I have some comments, most of them minor.

File: cmd/uhc/account/quota/cmd.go (hunk 12 above)
Comment (ghost) [reply cmd/uhc/account/quota/cmd.go:121]: Yes, It should never be more than one page

File: cmd/uhc/account/roles/cmd.go (hunk 17 above)
Comment (ghost) [reply cmd/uhc/account/roles/cmd.go]: I could not get the data to populate roles properly but this works just fine, I will fix comment

**Comment by ALimaRedHat** [comment]
[test]

**Comment by fahlmant** [comment]
[test]

**Comment by fahlmant** [comment]
[test]

**Comment by fahlmant** [comment]
[test]

**Comment by ALimaRedHat** [comment]
[test]

**Comment by fahlmant** [comment]
[test]

**Comment by jhernand** [comment]
[test]

**Review by jhernand — APPROVED** [review]

**Pull request merged**